UI-button: Make `ui_update` restore even after tab change / Page refresh

arturv2000 opened the pull request:
## Description

Needed a way to preserve the changes we send via `ui_update` even if the widget is hidden in another tab or after a page refresh.

As it was now, depending on the widget, it would only recuperate the last `ui_update` property sent. 
For the moment only did this for the `UI-Button` but if it is correct will make changes to at least `UI-Dropdown`, `UI-Slider` and `UI-Switch`, they are the widgets that at the moment I require this behaviour.

Changes:

- Unsubscribe from event 'widget-load:' to avoid multiple widget-Load messages, seems to solve #909
- When emitting a `widget-load` event, add to the message the `ui_update` parameters in store
- Add `ui_update.enable` as a valid parameter for `UI-Button`
- Add `onLoad` event to `UI-Button`
- Store the last message sent to  `UI-Button` to force a `widget-load` event upon refresh of page/widget.
    - This one, not quite sure about it. I needed a way for the `UI-Button` to trigger a `widget-load` event and this was the only way I could find to do that

This if applied to other widgets should also solve this issue #871



## Related Issue(s)

<!-- What issue does this PR relate to? -->

## Checklist

<!-- https://flowfuse.com/handbook/development/#defining-done -->

 - [ ] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass <!-- If not adding test coverage, please clarify why not? -->
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

joepavitt commented:
This _shouldn't_ be required... When sending the state on `widget-load`, we combine the "Dynamic Properties" (saved in the store) with the Node-RED config for the node, which should consequently load the relevant state as a combined entity

joepavitt commented:
I'm assuming this is because, currently `ui-button` is an action-driven widget, it doesn't maintain state. However, if we want to maintain a `enabled` state, we need to have this stored server-side to persist it?

arturv2000 commented:
> This _shouldn't_ be required... When sending the state on `widget-load`, we combine the "Dynamic Properties" (saved in the store) with the Node-RED config for the node, which should consequently load the relevant state as a combined entity

It isn't working either for `ui-switch` and `ui-dropdown` they were the three widgets I was testing to check the behaviour, the same was true for other properties, only the last `property` sent to the widget was being passed, since only the last message was being replayed.


> I'm assuming this is because, currently `ui-button` is an action-driven widget, it doesn't maintain state. However, if we want to maintain a `enabled` state, we need to have this stored server-side to persist it?

In this case (and most others i believe) the `enable` is not directly part of the `dynamic properties` it was only in the `props` field. I added to the `dynamic properties` and added also the store server-side in order to persist.

colinl commented:
> I'm assuming this is because, currently `ui-button` is an action-driven widget, it doesn't maintain state. However, if we want to maintain a `enabled` state, we need to have this stored server-side to persist it?

I believe that msg.enabled does work correctly with ui-button, and survives a refresh.  I am using it without problems.

arturv2000 commented:
> > I'm assuming this is because, currently `ui-button` is an action-driven widget, it doesn't maintain state. However, if we want to maintain a `enabled` state, we need to have this stored server-side to persist it?
> 
> I believe that msg.enabled does work correctly with ui-button, and survives a refresh. I am using it without problems.

The msg.enable survives a page refresh, but doesn't survive a page change (okay maybe not the best expression), it is in the case you are viewing Tab2 and send a msg.enable to a button (it also happens on other widgets and with other properties) on Tab1, when you return to tab1 the msg.enable is not taken in account, in other cases if you send multiple messages when the tab is not visible only the last message is replayed and only the properties that where updated in that last message are applied. 

It is the same situation as #871 but in this case for the msg.enable property.

There is an obvious solution to this, the user must keep track of all the properties / msg.ui_update that are sent to a widget, and check the event of tab change to `resend` it to the widget. But since we have a serve state store, we should make full use of it to simplify the process.

colinl commented:
> The msg.enable survives a page refresh, but doesn't survive a page change

It seems you are correct, though I cannot now understand why I don't see that in my 'production' system where it appears to work as it should, so I will have to investigate that.
In my nodes I maintain the complete state in the server state store, so that the client side gets the full state on refresh. When a message comes in I don't save the complete message in the state store, I just update the store with any data from the message. Then the message is passed to the client, and it uses that to update its copy of the server side state.
I have not yet worked out how to update the server state from the client, but so far that has not been necessary.  When I send a message from the client I always use the mode that does not update the store. I think I will have to implement a custom event on the socket to send an update to the server.

arturv2000 commented:
> In my nodes I maintain the complete state in the server state store, so that the client side gets the full state on refresh. When a message comes in I don't save the complete message in the state store, I just update the store with any data from the message. Then the message is passed to the client, and it uses that to update its copy of the server side state.


That is what I tried to do, that when a `widget-load` event is generated in `nodes/config/ui_base.js` to go the store and get the `ui_update` that are stored for that node

```js
msg.ui_update = statestore.getAll(id)
```
And send that message to the widget
And it was necessary to also include a `onLoad` method on the widget (Vue code) in order to process the message.


> 
> Edit: Just a note to say that I think the docs can be confusing in this area. When the browser is refreshed it does not resend the last message to the client, it sends the current contents of the server data store, which may not be the same thing.

Well, I think that is the latest message (sent or received, not quite sure, is one of that things that my brain finds confusing... too much habit of only writing/seeing C code), at least is what i could understand by what i saw on the websocket connection and some console.logs that I included in some points. 
From what i can understand access to all the content in the server store, is only made in `node.register` function, don't know exactly when it is called, and is passing the store contents to a 'state' variable.

colinl commented:
> Well, I think that is the latest message (sent or received, not quite sure

I am taking about third party nodes.  I think that in the core nodes the data store is always updated with the latest message in or out, and it seems to me that this is the cause of the problems.  In third party nodes it is up the author to decide how to handle the state store.  The technique I have used is to make sure that the data store always has the latest complete state information in it.

arturv2000 commented:
Close, since #1122 should handle this in a better / more organized way.
